### PR TITLE
Determine next log file number from directory on first run

### LINKED ISSUE #41326: W3CLogging file rotation seems to be deleting files out of order
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Use `dotnet new webapi` to create a new .NET 6 app on macOS or Linux. I haven't tested if this is happening on Windows too.

```
builder.Services.AddW3CLogging(options =>
{
    options.LogDirectory = "/olo/logs/http/loggingtest";
});
app.UseW3CLogging();
```

Restart the app 5 times. Note that each time the app restarts, a new file is created with an incrementing number starting with 0000.

After the fifth attempt, `w3clog-20220422.0000.txt` gets deleted, and `w3clog-20220422.0004.txt` gets created.

Restart the app another time. Note that this time, no new file is created, and logging stops happening for any future requests.

Setting a breakpoint in RollFiles indicates that the log counter is starting over at `0000`, but then immediately deleting the created file.  

Screen Shot 2022-04-22 at 3 03 25 PM

I'm happy to PR a fix if it's confirmed as a bug.

### Expected Behavior

File based w3c logging does not stop when the app restarts several times within the same day.

### Steps To Reproduce

Clone this repo and launch the app 6 times: https://github.com/dustinsoftware/logging-demo/blob/4780c75f3f69626382b809d306353d33461b328b/Program.cs#L9

### Exceptions (if any)

No exceptions or warnings that this is happening.

### .NET Version

6.0.101

### Anything else?

We noticed this issue on both macOS and Linux .NET 6 runtimes.

### PR DESCRIPTION
**Pull request by dustinsoftware**
# Determine next log file number from directory on first run

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description

Fixes #41326

### CODE REVIEW

**Comment by wtgodbe** [comment]
Could you explain how this is different from the current logic? We already attempt to pick up where we left off on restart:

https://github.com/dotnet/aspnetcore/blob/bcdb13d1f2e42f2d318ead75e5e86fc27c78a117/src/Middleware/HttpLogging/src/FileLoggerProcessor.cs#L151-L171

There is a bug on line 162, though, which should be `if (_fileNumber >= W3CLoggerOptions.MaxFileCount - 1)`

**Comment by dustinsoftware** [comment]
I added a test in [FileLoggerProcessorTests.cs](https://github.com/dotnet/aspnetcore/pull/41375/files#diff-eb77a779f77a9a29a358efd0b9aedeb6941383ce8b25a701da4bf9e868c7b433) to clarify what this logic is intended to address. If you stick a breakpoint on this line, the fileCounter is now starting at 4 when the processor is constructed again. Previously it would start back over at 0. 

https://github.com/dotnet/aspnetcore/pull/41375/files#diff-eb77a779f77a9a29a358efd0b9aedeb6941383ce8b25a701da4bf9e868c7b433R395

> There is a bug on line 162, though, which should be if (_fileNumber >= W3CLoggerOptions.MaxFileCount - 1)

I'll fix that and add a test.

File: src/Middleware/HttpLogging/test/FileLoggerProcessorTests.cs [hunk 1]
```diff
@@ -349,6 +350,72 @@ await using (var logger = new FileLoggerProcessor(new OptionsWrapperMonitor<W3CL
             Helpers.DisposeDirectory(path);
         }
     }
+    [Fact]
+    public async Task RollsTextFiilesWhenFirstLogOfDayIsMissing()
```
Comment (wtgodbe) [inline src/Middleware/HttpLogging/test/FileLoggerProcessorTests.cs:354]:
```suggestion
    public async Task RollsTextFilesWhenFirstLogOfDayIsMissing()
```

**Review by wtgodbe — APPROVED** [review]
I see, seems like a good behavioral change. Thanks for the catch & fix!

**Comment by dustinsoftware** [comment]
Tests are updated, let me know if there's anything missing. 

With this PR, the assignment to `_fileNumber` inside the `_firstFile` block should always be the number of the (not yet written) new log file, and if the value is `10000` we know the folder has been filled up, so I think as-is the check is correct. There is now a test asserting that is the case for both the app-running and app-restarted cases.

**Comment by dustinsoftware** [comment]
Would it be possible to backport this fix into net6 since that is LTS?

**Comment by wtgodbe** [comment]
> Would it be possible to backport this fix into net6 since that is LTS?

Generally we hesitate to take behavioral changes back to servicing since somebody might be relying on the old behavior, but in this case that seems unlikely. I'll open a backport of this against the release/6.0 branch & see if we can get it through the review process.

**Comment by wtgodbe** [comment]
Ah, right, file-scoped-namespaces. @dustinsoftware could you open a backport of this against the `release/6.0` branch, and I'll handle getting it through shiproom review? Branches for servicing are opening tomorrow thru next Monday, so if we can get it in by then it'll be in 6.0 for the June release.

**Comment by BrennanConroy** [comment]
> Ah, right, file-scoped-namespaces

Thought that was fixed with https://github.com/dotnet/aspnetcore/pull/41087?

**Comment by wtgodbe** [comment]
> Thought that was fixed with https://github.com/dotnet/aspnetcore/pull/41087?

I thought that only affected `git blame`

**Comment by dustinsoftware** [comment]
>  @dustinsoftware could you open a backport of this against the release/6.0 branch, and I'll handle getting it through shiproom review

sure :)

**Pull request merged**